python/tvm/relay/qnn/op/qnn.py

```diff
@@ -72,3 +74,120 @@ def requantize(data,
                             output_zero_point,
                             rounding,
                             out_dtype)
+
+
+def add(lhs, rhs, lhs_scale, lhs_zero_point, rhs_scale, rhs_zero_point, output_scale,
+        output_zero_point):
+    """Quantized addition with numpy-style broadcasting.
+
+    Parameters
+    ----------
+    lhs : relay.Expr
+        The left hand side quantized input data.
+
+    rhs : relay.Expr
+        The right hand side quantized input data.
+
+    lhs_scale: float
+        The scale of the lhs quantized expr.
+
+    lhs_zero_point: int
+       The zero point of lhs quantized expr.
+
+    rhs_scale: float
+        The scale of the rhs quantized expr.
+
+    rhs_zero_point: int
+       The zero point of rhs quantized expr.
+
+    output_scale: float
+        The scale of the output quantized expr.
+
+    output_zero_point: int
+       The zero point of output quantized expr.
+
+    Returns
+    -------
+    result : relay.Expr
+        The computed result.
+
+    """
+
+    # Find the dtype of the input expr. This is required for the requantize op. Since, this is
+    # add op, the dtype of the input is same as dtype of the output.
+    data0 = relay.transform.infer_type(lhs)
+    in_dtype = data0.checked_type.dtype
+    assert in_dtype in ('int8', 'uint8')
+
+    # First, check if the qnn params of lhs and rhs match. If yes, we can avoid one requantize by
+    # calling add first and then requantize. The whole process can be represented as follows
+    #
+    #          scale_c * (Q_c - zp_c) = scale * (Q_a - zp) + scale * (Q_b - zp)
+    #          scale_c * (Q_c - zp_c) = scale * (Q_a + Q_b - zp - zp)
+    #
+    # RHS looks like a quantized tensor with scale = scale, and zero_point = zp + zp
+    # This can be handled by first subtracting the zero point, followed by requantize with the
+    # output qnn params. The add op is done in int32 precision.
+
+    if lhs_scale == rhs_scale and lhs_zero_point == rhs_zero_point:
+        lhs = relay.cast(lhs, dtype='int32')
```

Comment:
add/sub in int16 should be enough, but not a big deal :)

Reply:
Tried that :) Currently, it fails before requantize input can only be (uint8, int8, int32). I think for now, it should be ok. If we see more demand of int16, we can add support across all the QNN ops.

Reply:
Yes, that is far from a blocking issue :) Thank you.